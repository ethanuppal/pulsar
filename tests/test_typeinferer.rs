#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use pulsar::{
        frontend::{infer::TypeInferer, lexer::Lexer, parser::Parser},
        utils::{error::ErrorManager, loc::Source}
    };
    use std::{cell::RefCell, fs, rc::Rc};

    fn read(filename: &str) -> Rc<Source> {
        Source::file(
            filename.into(),
            fs::read_to_string(filename)
                .expect(format!("Could not read file: {}", filename).as_str())
        )
    }

    fn typeinferer_output(
        filename: &str, error_manager: Rc<RefCell<ErrorManager>>
    ) -> String {
        let source = read(filename);
        let lexer = Lexer::new(source, error_manager.clone());
        let tokens: Vec<_> = lexer.into_iter().collect();
        let parser = Parser::new(tokens, error_manager.clone());
        let program_ast: Vec<_> = parser.into_iter().collect();
        assert_eq!(false, error_manager.borrow().has_errors());
        let mut type_inferer = TypeInferer::new(error_manager.clone());
        let annotated_ast_opt = type_inferer.infer(program_ast);

        let mut output = String::new();
        if let Some(annotated_ast) = annotated_ast_opt {
            for node in annotated_ast {
                output.push_str(node.to_string().as_str());
            }
        }
        let mut buffer = Vec::new();
        if error_manager.borrow().has_errors() {
            error_manager
                .borrow_mut()
                .consume_and_write(&mut buffer)
                .unwrap_or_default();
        }
        output.push_str(String::from_utf8(buffer).unwrap().as_str());
        output
    }

    #[test]
    fn test_typeinferer() {
        let error_manager = ErrorManager::with_max_count(5);
        assert_snapshot!(typeinferer_output(
            "tests/data/infer1.pl",
            error_manager.clone()
        ));

        assert_snapshot!(typeinferer_output(
            "tests/data/infer2.pl",
            error_manager.clone()
        ));

        assert_snapshot!(typeinferer_output(
            "tests/data/infer3.pl",
            error_manager.clone()
        ));

        assert_snapshot!(typeinferer_output(
            "tests/data/infer4.pl",
            error_manager.clone()
        ));
    }
}
