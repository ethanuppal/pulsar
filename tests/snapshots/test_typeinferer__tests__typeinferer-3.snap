---
source: tests/test_typeinferer.rs
expression: "typeinferer_output(\"tests/data/infer3.pl\", error_manager.clone())"
---
func main() -> Unit {
    let big: Int64[10] = [...]: Int64[1]
    let small: Int64[1] = big: Int64[1]
}error[E0010]: tests/data/infer3.pl:2:9: Array sizes don't match: 10 != 1
   1 │  func main() {
   2 │      let big: Int[10] = [...]
   3 │      let small: Int[1] = big
...
   1 │  func main() {
   2 │      let big: Int[10] = [...]
                               │ Inferred to have size 1 here based on environment
   3 │      let small: Int[1] = big
